x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/randomsampling/RandomSamplingQuery.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.randomsampling;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.ConjunctionUtils;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.xpack.ml.aggs.FastGeometric;
+import org.elasticsearch.xpack.ml.aggs.PCG;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.IntSupplier;
+
+/**
+ * A query that randomly matches documents with a user-provided probability within a geometric distribution
+ */
+public final class RandomSamplingQuery extends Query {
+
+    private final double p;
+    private final boolean cacheable;
+    private final Query query;
+    private final int seed;
+    private final long hash;
+
+    /**
+     * @param p         The sampling probability e.g. 0.05 == 5% probability a document will match
+     * @param seed      The seed from the builder
+     * @param hash      A unique hash so that if the same seed is used between multiple queries, unique random number streams
+     *                  can be generated
+     * @param cacheable True if the seed is static (provided by the user) so that we can cache the query, false otherwise
+     */
+    public RandomSamplingQuery(double p, int seed, long hash, boolean cacheable, Query query) {
```

Comment:
I wouldn't take the `boolean cacheable` and let Lucene decide whether it sees this query over and over or not for simplicity.